pkg/dns/vips_allocator.go

```diff
@@ -199,63 +265,113 @@ func addFromMeshGateway(outboundSet *vips.VirtualOutboundMeshView, dnsSuffix, me
-	outboundSet := vips.NewEmptyVirtualOutboundView()
+func (d *VIPsAllocator) fetchZoneIngresses(ctx context.Context) ([]*core_mesh.ZoneIngressResource, error) {
+	zoneIngresses := core_mesh.ZoneIngressResourceList{}
+	if err := d.rm.List(ctx, &zoneIngresses); err != nil {
+		return nil, err
+	}
+
+	return zoneIngresses.Items, nil
+}
+
+func (d *VIPsAllocator) fetchDataplanesByMesh(ctx context.Context) (map[string][]*core_mesh.DataplaneResource, error) {
+	dataplanes := core_mesh.DataplaneResourceList{}
+	if err := d.rm.List(ctx, &dataplanes); err != nil {
+		return nil, err
+	}
+
+	out := map[string][]*core_mesh.DataplaneResource{}
+	for _, dataplane := range dataplanes.Items {
+		out[dataplane.Meta.GetMesh()] = append(out[dataplane.Meta.GetMesh()], dataplane)
+	}
 
+	return out, nil
+}
+
+func (d *VIPsAllocator) fetchVirtualOutboundsByMesh(ctx context.Context) (map[string][]*core_mesh.VirtualOutboundResource, error) {
 	virtualOutbounds := core_mesh.VirtualOutboundResourceList{}
-	if err := d.rm.List(ctx, &virtualOutbounds, store.ListByMesh(mesh)); err != nil {
+	if err := d.rm.List(ctx, &virtualOutbounds); err != nil {
 		return nil, err
 	}
-	dataplanes := core_mesh.DataplaneResourceList{}
-	if err := d.rm.List(ctx, &dataplanes, store.ListByMesh(mesh)); err != nil {
+
+	out := map[string][]*core_mesh.VirtualOutboundResource{}
+	for _, virtualOutbound := range virtualOutbounds.Items {
+		mesh := virtualOutbound.Meta.GetMesh()
+		out[mesh] = append(out[mesh], virtualOutbound)
+	}
+
+	return out, nil
+}
+
+func (d *VIPsAllocator) fetchExternalServicesByMesh(ctx context.Context) (map[string][]*core_mesh.ExternalServiceResource, error) {
+	externalServices := core_mesh.ExternalServiceResourceList{}
+	if err := d.rm.List(ctx, &externalServices); err != nil {
 		return nil, err
 	}
 
+	out := map[string][]*core_mesh.ExternalServiceResource{}
+	for _, externalService := range externalServices.Items {
+		mesh := externalService.Meta.GetMesh()
+		out[mesh] = append(out[mesh], externalService)
+	}
+
+	return out, nil
+}
+
+func (d *VIPsAllocator) fetchMeshGatewaysByMesh(ctx context.Context) (map[string][]*core_mesh.MeshGatewayResource, error) {
+	meshGateways := core_mesh.MeshGatewayResourceList{}
+	if err := d.rm.List(ctx, &meshGateways); err != nil {
+		return nil, err
+	}
+
+	out := map[string][]*core_mesh.MeshGatewayResource{}
+	for _, meshGateway := range meshGateways.Items {
+		mesh := meshGateway.Meta.GetMesh()
+		out[mesh] = append(out[mesh], meshGateway)
+	}
+
+	return out, nil
+}
+
+func (d *VIPsAllocator) buildVirtualOutboundMeshView(mesh string, virtualOutbounds []*core_mesh.VirtualOutboundResource, dataplanes []*core_mesh.DataplaneResource, zoneIngresses []*core_mesh.ZoneIngressResource, externalServices []*core_mesh.ExternalServiceResource, meshGatewaysByMesh map[string][]*core_mesh.MeshGatewayResource) (*vips.VirtualOutboundMeshView, error) {
```

Comment:
That's a lot of parameters. We could define a struct holding them if you think that's necessary for lisibility

Reply:
it's a private function. It's not ideal, but I think it's fine. The only thing is that I'd format this in this way
```
func (d *VIPsAllocator) buildVirtualOutboundMeshView(
  mesh string,
  virtualOutbounds []*core_mesh.VirtualOutboundResource,
  dataplanes []*core_mesh.DataplaneResource,
  zoneIngresses []*core_mesh.ZoneIngressResource,
  externalServices []*core_mesh.ExternalServiceResource,
  meshGatewaysByMesh map[string][]*core_mesh.MeshGatewayResource,
) (*vips.VirtualOutboundMeshView, error) {
```

Reply:
🤝 done